go/auth/token.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+//
+// Code used to support authentication tokens for arbitrary purposes.
+//
+package auth
+
+import (
+	"time"
+
+	libkb "github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol"
+	jsonw "github.com/keybase/go-jsonw"
+)
+
+const CurrentTokenVersion = 1
+
+type Token struct {
+	User          keybase1.UID
+	Username      libkb.NormalizedUsername
+	Key           keybase1.KID
+	TokenType     string
+	CreationTime  int64
+	ExpireIn      int
+	Tag           string
+	Version       int
+	ClientName    string
+	ClientVersion string
+}
+
+func NewToken(user keybase1.UID, username libkb.NormalizedUsername, key keybase1.KID,
+	tokenType string, expireIn int, clientName string, clientVersion string) *Token {
+	return &Token{
+		User:          user,
+		Username:      username,
+		Key:           key,
+		TokenType:     tokenType,
+		CreationTime:  time.Now().Unix(),
+		ExpireIn:      expireIn,
+		ClientName:    clientName,
+		ClientVersion: clientVersion,
+	}
+}
+
+func ParseToken(token string) (*Token, error) {
+	jw, err := jsonw.Unmarshal([]byte(token))
+	if err != nil {
+		return nil, err
+	}
+	t := &Token{}
+	if err := t.fromJSON(jw); err != nil {
+		return nil, err
+	}
+	return t, nil
+}
+
+func (arg Token) Bytes() []byte {
+	var err error
+	var tmp []byte
+	if tmp, err = arg.toJSON().Marshal(); err != nil {
+		return []byte{}
+	}
+	return tmp
+}
+
+func (arg Token) String() string {
+	return string(arg.Bytes())
+}
+
+func (arg Token) Verify(signature, tokenType string, maxExpireIn int) error {
+	key, err := libkb.ImportKeypairFromKID(arg.Key)
+	if err != nil {
+		return err
+	}
+	if _, err := key.VerifyString(signature, arg.Bytes()); err != nil {
+		return err
+	}
+	if tokenType != arg.TokenType {
+		return InvalidTokenTypeError{
+			ExpectedTokenType: tokenType,
+			ReceivedTokenType: arg.TokenType,
+		}
+	}
+	remaining := arg.TimeRemaining()
+	if remaining > maxExpireIn {
+		return MaxTokenExpiresError{
+			CreationTime: arg.CreationTime,
+			ExpireIn:     arg.ExpireIn,
+			Now:          time.Now().Unix(),
+			MaxExpireIn:  maxExpireIn,
+			Remaining:    remaining,
+		}
+	}
+	if remaining <= 0 {
+		return TokenExpiredError{
+			CreationTime: arg.CreationTime,
+			ExpireIn:     arg.ExpireIn,
+			Now:          time.Now().Unix(),
+		}
+	}
+	return nil
+}
+
+func (arg Token) TimeRemaining() int {
+	now := time.Now().Unix()
```

Comment:
As long as we're being more virtuous (see above), maybe we ought to use Go's `time.Time` and `time.Duration` classes to do time manipulation